rfcs/2023-react-aria-components.md

```diff
@@ -0,0 +1,330 @@
+
+* Components – documentation for the new component API, with sub-sections for each category as we have today (e.g. buttons, collections, date and time, etc.).
+* Hooks – our existing hook documentation for ARIA patterns, without the hooks moved into the below sections.
+* Patterns – hook documentation for lower level patterns that span multiple ARIA patterns, such as collections, selection, virtualizer, drag and drop, etc.
+* Interactions – documentation for interactions like usePress, useHover, FocusScope, etc.
+* Utilities – documentation for various utilities that don't fit into the other categories, like i18n, SSR, mergeProps, etc.
+
+Each component page should have several sections, similar to the documentation for React Spectrum:
+
+* A main example showing a minimal real-world scenario. The CSS for this should be collapsed underneath so it can be expanded only if needed.
+* Features – a list of the main features of the component, grouped by category to allow easy skimming.
+* Anatomy – the anatomy diagram and description of the various parts of the component. For components that compose other child components, links to the documentation for those components would be useful, along with any conceptual documentation (e.g. collections, selection).
+* Prop table for all relevant components.
+* Styling – a description of how to style the component in various ways, including documentation of all of the available states and CSS selectors. Links to styled codesandbox examples with various tools like we have in our hook docs might be useful as well.
+* Reusable wrappers – an example of creating a reusable component with the styling implementation and all of the parts included.
+* Usage – the usage section copied from the hook documentation.
+* Advanced customization – if applicable, shows how to mix and match this component with the hooks and links to all of the relevant hook documentation for when more customization is needed.
+
+This may seem like a lot, but most of this can either be copied from existing React Aria hook documentation, or reused from other pages (e.g. most of the styling guide).
+
+Overall, the documentation should aim to show as many real-world examples as possible. The examples should also be styled to look somewhat pretty, rather than the bare-bones styles in our current React Aria docs. The example styles can leave a significant impression on readers even if styling isn't the point, so we want to make them look good to show what is possible. This includes things like supporting light and dark mode, windows high contrast mode, etc. People can also copy and paste the default styles as a starting point, so it is good to capture as many cases as possible.
```

Comment:
Love the proposal overall! This would increase visibility for such a crazy powerful library. 

I'd like to emphasize this paragraph in particular,
> **_documentation should aim to show as many real-world examples as possible. The examples should also be styled to look somewhat pretty_**

I think that change alone could increase the adoption of the current aria hooks

---- 

Different topic: 

So much flexibility could potentially open the door for more user error (i.e. passing an improperly structured custom component to a ListBox), I wonder if OOTB we could add [ErrorBoundaries](https://reactjs.org/docs/error-boundaries.html) to each of the building blocks thus narrowing down the error to the specific component instead of crashing the app ... ErrorBoundaries + TS type-safety could potentially improve the dev experience preventing and/or catching errors gracefully, what do you think?

Reply:
Not sure about error boundaries being built into the components – what would we render? But we will put some effort into providing good error messages/warnings that help the developer know when they put invalid children in a component (e.g. interactivity where it is not allowed by the ARIA specs). I think in many cases those would also be warnings rather than errors anyway.